### PR TITLE
concentration parameter for document and topics distributions in ml_lda v2

### PR DESCRIPTION
**Pull request by MarcinKosinski**

### CODE REVIEW

**Comment by MarcinKosinski** [comment]
@kevinushey do you think we could add some tests after merging this request to extend ml_lda : ) ?

**Comment by MarcinKosinski** [comment]
According to original scala parameters https://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.ml.clustering.LDA

````{bash}

final val topicConcentration: DoubleParam

Concentration parameter (commonly named "beta" or "eta") for the prior placed on topics' distributions over terms.

This is the parameter to a symmetric Dirichlet distribution.

Note: The topics' distributions over terms are called "beta" in the original LDA paper by Blei et al., but are called "phi" in many later papers such as Asuncion et al., 2009.

If not set by the user, then topicConcentration is set automatically. (default = automatic)

Optimizer-specific parameter settings:

    EM
        Value should be > 1.0
        default = 0.1 + 1, where 0.1 gives a small amount of smoothing and +1 follows Asuncion et al. (2009), who recommend a +1 adjustment for EM.
    Online
        Value should be >= 0
        default = (1.0 / k), following the implementation from https://github.com/Blei-Lab/onlineldavb.

Definition Classes
    LDAParams
Annotations
    @Since( "1.6.0" ) 




 final val topicDistributionCol: Param[String]

Output column with estimates of the topic mixture distribution for each document (often called "theta" in the literature). Returns a vector of zeros for an empty document.

This uses a variational approximation following Hoffman et al. (2010), where the approximate distribution is called "gamma." Technically, this method returns this approximation "gamma" for each document.

Definition Classes
    LDAParams
Annotations
    @Since( "1.6.0" ) 
````

**Comment by javierluraschi** [comment]
@kevinushey merge before CRAN checkpoint? Thoughts? BTW. I'm not merging any PR at this point, I was thinking we can get together next week with JJ and cherry pick the safe ones.

**Comment by kevinushey** [comment]
I think this should be fine to merge as it doesn't lie in any of the Spark critical paths.

File: R/ml_lda.R [hunk 1]
```diff
@@ -7,16 +7,35 @@
...
 #' @export
 ml_lda <- function(x,
                    features = dplyr::tbl_vars(x),
                    k = length(features),
-                   ml.options = ml_options(),
-                   ...)
-{
+                   alpha = (50 / k) + 1,
+                   beta = 0.1 + 1,
+                   ml.options = NULL,
```
Comment (kevinushey) [inline R/ml_lda.R]: This should be `ml.options = ml_options()` just to ensure the parameter is non-null.

Comment (MarcinKosinski) [inline R/ml_lda.R]: @kevinushey I've updated this option with https://github.com/rstudio/sparklyr/pull/186/commits/cbc0656c1c1c7fef1ae1c94081a14564c19c7f71 thanks for the suggestion

**Comment by kevinushey** [comment]
Thanks!

**Comment by dsblr** [comment]
@MarcinKosinski I am new to sparklyr. I have implemented a small example code do LDA using R and like to replicate the same using sparklyr. But I am not getting any clue, how to proceed further. Please help me with the steps to replicate the following R code to Sparklyr.

data1 <- read.csv(system.file("data/NYTimes.csv.gz",package="maxent"))
##create DTM
matrix <- create_matrix(cbind.data.frame(data1$Title,data1$Subject), language="english", removeNumbers=TRUE, stemWords=TRUE, weighting=weightTf)

##Perform Latent Dirichlet Allocation 
##First determine the number of topics in our data.
k <- length(unique(data1$Topic.Code))        ## Here the value of K is 27 as there are 27 different topics the author has identified
lda <- LDA(matrix, k,method = "VEM")

I am unable to create DTM from the corpus in sparklyr; which I think the input for ml_lda. Please help me with the steps.
sc <- spark_connect(master = "local")
sdf <- spark_read_csv(sc = sc, name = "tbl1",system.file("data/NYTimes.csv.gz",package="maxent"))
sdf1 <- spark_dataframe(sdf)
tokenized <-  sdf1 %>% 
  na.omit() %>%
  ft_tokenizer(input.col = "Subject", output.col = "tokens")

**Comment by MarcinKosinski** [comment]
@dsblr I need your whole code like:

- including packages, so that I know from which package you take `LDA` or `create_matrix` function
- session info, obtained by `sessionInfo()`, so that I am aware of packages version numbers with which you work - you might have old versions with bugs

By the way: are you sure you are using `sparklyr` and not `SparkR` package? I don't know the `ft_tokenizer` function.

**Comment by dsblr** [comment]
@MarcinKosinski Thanks for your response. Please find my response below

**LDA topic modeling using R packages**
sessioninfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows Server >= 2012 x64 (build 9200)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] topicmodels_0.2-4 RTextTools_1.4.2  maxent_1.3.3.1    SparseM_1.74      tm_0.6-2          NLP_0.1-9        

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.7         base64enc_0.1-3     bitops_1.0-6        class_7.3-14        iterators_1.0.8     tools_3.3.1         rpart_4.1-10        digest_0.6.10       ipred_0.9-5        
[10] jsonlite_1.1        tibble_1.2          lattice_0.20-33     Matrix_1.2-6        foreach_1.4.3       DBI_0.5-1           parallel_3.3.1      prodlim_1.5.7       e1071_1.6-7        
[19] dplyr_0.5.0         httr_1.2.1          caTools_1.17.1      stats4_3.3.1        rprojroot_1.1       nnet_7.3-12         grid_3.3.1          glmnet_2.0-5        tree_1.0-37        
[28] tau_0.0-18          R6_2.2.0            survival_2.39-4     lava_1.4.6          magrittr_1.5        backports_1.0.4     codetools_0.2-14    modeltools_0.2-21   MASS_7.3-45        
[37] splines_3.3.1       randomForest_4.6-12 assertthat_0.1      slam_0.1-40    

data1 <- read.csv(system.file("data/NYTimes.csv.gz",package="maxent"))
 ##create DTM
 matrix <- create_matrix(cbind.data.frame(data1$Title,data1$Subject), language="english", removeNumbers=TRUE, stemWords=TRUE, weighting=weightTf)

##Perform Latent Dirichlet Allocation
 ##First determine the number of topics in our data.
 k <- length(unique(data1$Topic.Code)) ## Here the value of K is 27 as there are 27 different topics the author has identified
 lda <- LDA(matrix, k,method = "VEM")


**Would like to replicate the same using sparklyr ml_lda**

**LDA topic modeling using sparklyr**

sessioninfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows Server >= 2012 x64 (build 9200)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_0.5.0     sparklyr_0.4.41

loaded via a namespace (and not attached):
 [1] httr_1.2.1      backports_1.0.4 magrittr_1.5    R6_2.2.0        assertthat_0.1  rprojroot_1.1   DBI_0.5-1       tools_3.3.1     base64enc_0.1-3 tibble_1.2      Rcpp_0.12.7    
[12] jsonlite_1.1    digest_0.6.10  

partial code which I am referring for tokenization is availed on https://github.com/rstudio/sparklyr/issues/279#issuecomment-271640924 . 
ft_tokenizer is not listed as a package in the web page of sparklyr, 
but you can find this in your sparklyr session; sparklyr::ft_tokenizer()

sc <- spark_connect(master = "local")
 sdf <- spark_read_csv(sc = sc, name = "tbl1",system.file("data/NYTimes.csv.gz",package="maxent"))
 sdf1 <- spark_dataframe(sdf)
 tokenized <- sdf1 %>%
 na.omit() %>%
 ft_tokenizer(input.col = "Subject", output.col = "tokens")

I am wondering that you have used ml_lda, if yes please help me with the steps to create DTM/TDM from a corpus and pass this to ml_lda. Please help me with the alternative if anything available for topic modeling in LDA using sparklyr.

Thank you for looking into this.

**Comment by MarcinKosinski** [comment]
@dsblr which `libraries` do you use with `library`? So that I will be able to copy-paste your code without any guesses?

Have you tried then to simply `ml_lda(tokenized)` or `ml_lda(sf1)`?

**Comment by dsblr** [comment]
@MarcinKosinski  thanks for your response. I have just updated the libraries in my previous comment.
for sparklyr I used library(sparklyr) and library(dplyr).

I am getting the following error when I tried to execute ml_lda(tokenized) or ml_lda(sdf1)

> ml_lda(tokenized)
* No rows dropped by 'na.omit' call
Error: java.lang.IllegalArgumentException: Data type StringType is not supported.
	at org.apache.spark.ml.feature.VectorAssembler$$anonfun$transformSchema$1.apply(VectorAssembler.scala:121)
	at org.apache.spark.ml.feature.VectorAssembler$$anonfun$transformSchema$1.apply(VectorAssembler.scala:117)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at org.apache.spark.ml.feature.VectorAssembler.transformSchema(VectorAssembler.scala:117)
	at org.apache.spark.ml.PipelineStage.transformSchema(Pipeline.scala:70)
	at org.apache.spark.ml.feature.VectorAssembler.transform(VectorAssembler.scala:54)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at sparklyr.Invoke$.invoke(invoke.scala:94)
	at sparklyr.StreamHandler$.handleMethodCall(stream.scala:89)
	at sparklyr.StreamHandler$.read(stream.scala:55)
	at sparklyr.BackendHandler.channelRead0(handler.scala:49)
	at sparklyr.BackendHandler.channelRead0(handler.scala:14)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:244)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:846)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:131)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:511)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:111)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
	at java.lang.Thread.run(Thread.java:745)
> ml_lda(sdf)
* Dropped 15 rows with 'na.omit' (100 => 85)
Error: java.lang.IllegalArgumentException: Data type StringType is not supported.
	at org.apache.spark.ml.feature.VectorAssembler$$anonfun$transformSchema$1.apply(VectorAssembler.scala:121)
	at org.apache.spark.ml.feature.VectorAssembler$$anonfun$transformSchema$1.apply(VectorAssembler.scala:117)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at org.apache.spark.ml.feature.VectorAssembler.transformSchema(VectorAssembler.scala:117)
	at org.apache.spark.ml.PipelineStage.transformSchema(Pipeline.scala:70)
	at org.apache.spark.ml.feature.VectorAssembler.transform(VectorAssembler.scala:54)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at sparklyr.Invoke$.invoke(invoke.scala:94)
	at sparklyr.StreamHandler$.handleMethodCall(stream.scala:89)
	at sparklyr.StreamHandler$.read(stream.scala:55)
	at sparklyr.BackendHandler.channelRead0(handler.scala:49)
	at sparklyr.BackendHandler.channelRead0(handler.scala:14)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:244)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:846)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:131)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:511)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:111)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
	at java.lang.Thread.run(Thread.java:745)

**Comment by iamkbpark** [comment]
@dsblr @MarcinKosinski   I am also having the same error messages for both `ml_lda(tokenized)` & `ml_lda(sdf1)`, and looking forward to hearing some tips. Thank you for your help!

**Comment by MarcinKosinski** [comment]
I have no idea how to proceed with those tokens from `ft_tokenized` but I
will think about this.
I think the input into `ml_lda` should be a something like
DocumentTermMatrix where
`tokenized` has this information in `tokens` column but in a list format.

Extracting `tokens` column and transferring it to a `something like
DocumentTermMatrix` should be a valid input into a `ml_lda`


2017-01-12 7:00 GMT+01:00 iamkbpark <notifications@github.com>:

**Comment by dsblr** [comment]
@MarcinKosinski thanks for your response. I am unable to convert tokens to DTM, for me its complex. Could you please help me with the srating steps.

Alternatively I am planning create udf in scala for TF-IDF and expose that to sparklyr environment.

**Comment by dsblr** [comment]
@MarcinKosinski could you please help me to understand how the following code works ?
invoke_new(sc, "org.apache.spark.ml.feature.HashingTF") 

I would like to know what is the path(physical path or OS path) of the org.apache.spark.ml.feature.HashingTF  ?

**Comment by MarcinKosinski** [comment]
@dsblr I don't know. Actually I am not in the RStudio team. Have you
checked sparklyr website? http://spark.rstudio.com/
I haven't been using sparklyr for 2 or 3 monts, but now I see that the
website is full of examples and description.

You can also check my blog post about extending sparklyr
http://r-addict.com/2016/08/25/Extending-Sparklyr.html

2017-01-13 6:10 GMT+01:00 dsblr <notifications@github.com>:

**Pull request merged**